pkg/kv/kvtest/store.go

```diff
@@ -369,3 +374,333 @@ func testScanPrefix(t *testing.T, ms MakeStore) {
+
+	// scan and delete the same prefix
+	testDeleteWhileIterSamePrefixSingleRun(t, ms, allPrefs, ininPref, ininPref)
+}
+
+func testDeleteWhileIterSamePrefixSingleRun(t *testing.T, ms MakeStore, prefsToCreate [][]byte, readPref, delPref []byte) {
+	ctx := context.Background()
+	store := ms(t, ctx)
+	defer store.Close()
+
+	// Emptying the KV store from previous entries. This is essential in these tests as
+	// the verification relies on exact keys to exist
+	cleanIter, err := store.Scan(ctx, delPref)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer cleanIter.Close()
+
+	keysToClean := [][]byte{}
+	for cleanIter.Next() {
+		e := cleanIter.Entry()
+		if e != nil {
+			keysToClean = append(keysToClean, e.Key)
+		} else {
+			t.Fatal("unexpected nil entry during pre test clean", cleanIter.Err())
+		}
+	}
+
+	for _, k := range keysToClean {
+		err = store.Delete(ctx, k)
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	// originalKv holds all the created items, to later verify that no items, that do not fit delPref, were deleted
+	originalKv := []kv.Entry{}
+	for _, pref := range prefsToCreate {
+		originalKv = append(originalKv, setupSampleData(t, ctx, store, string(pref), 100)...)
+	}
+
+	// Will be filled by the scan&read routine, to later verify that all values are passed by scan
+	readDone := make(map[string]bool)
+	deleteDone := make(map[string]bool)
+
+	errGrp, errCtx := errgroup.WithContext(context.Background())
+
+	// Scan and read
+	errGrp.Go(func() error {
+		ei, err := store.Scan(errCtx, readPref)
+		if err != nil {
+			return fmt.Errorf("unexpected error from store.Scan (read): %w", err)
+		}
+		defer ei.Close()
+
+		for ei.Next() {
+			e := ei.Entry()
+			if e != nil {
+				readDone[string(e.Key)] = true
+			} else if ei.Err() != nil {
+				return fmt.Errorf("unexpected error during read iteration: %w", err)
+			}
+		}
+		return nil
+	})
+
+	// Scan and delete
+	errGrp.Go(func() error {
+		ei, err := store.Scan(errCtx, delPref)
+		if err != nil {
+			return fmt.Errorf("unexpected error from store.Scan (delete): %w", err)
+		}
+		defer ei.Close()
+
+		for ei.Next() {
+			e := ei.Entry()
+			if e == nil {
+				return fmt.Errorf("unexpected nil entry during deletion interation: %w", ei.Err())
+			}
+			err = store.Delete(errCtx, e.Key)
+			if err != nil {
+				return fmt.Errorf("unexpected delete failure :%w", err)
+			}
+			deleteDone[string(e.Key)] = true
+		}
+		return nil
+	})
+
+	err = errGrp.Wait()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// verify all readDone entries fits readPref
+	for k := range readDone {
+		if strings.Index(k, string(readPref)) != 0 {
+			t.Fatal("unexpected key read for prefix", k, string(readPref))
+		}
+	}
+
+	// verify all delete entries fits delPref
+	for k := range deleteDone {
+		if strings.Index(k, string(delPref)) != 0 {
+			t.Fatal("unexpected key deleted for prefix", k, string(delPref))
+		}
+	}
+
+	// verify entries that does not fit delPref, were not accidentally deleted
```

Comment:
Seems like many of the validations on both tests can be exported to a helper function to reduce code duplication

Reply:
Done